trie/trie.go

```diff
@@ -50,230 +50,166 @@ func (self *Trie) Iterator() *Iterator {
+		if tn == nil {
+			return nil
+		}
+		switch n := tn.(type) {
+		case shortNode:
+			if len(key) < len(n.Key) || !bytes.Equal(n.Key, key[:len(n.Key)]) {
+				return nil
+			}
+			tn = n.Val
+			key = key[len(n.Key):]
+		case fullNode:
+			tn = n[key[0]]
+			key = key[1:]
+		case hashNode:
+			tn = self.resolveHash(n)
+		default:
+			panic(fmt.Sprintf("%T: invalid node: %v", tn, tn))
+		}
+	}
+	return tn
 }
 
-func (self *Trie) insert(node Node, key []byte, value Node) Node {
+func (self *Trie) insert(n node, key []byte, value node) node {
 	if len(key) == 0 {
 		return value
 	}
-
-	if node == nil {
-		node := NewShortNode(self, key, value)
-		node.dirty = true
-		return node
+	if n == nil {
+		return shortNode{key, value}
 	}
-
-	switch node := node.(type) {
-	case *ShortNode:
-		k := node.Key()
-		cnode := node.Value()
-		if bytes.Equal(k, key) {
-			node := NewShortNode(self, key, value)
-			node.dirty = true
-			return node
-
+	switch n := n.(type) {
+	case shortNode:
+		if bytes.Equal(n.Key, key) {
+			return shortNode{key, value}
 		}
-
-		var n Node
-		matchlength := MatchingNibbleLength(key, k)
-		if matchlength == len(k) {
-			n = self.insert(cnode, key[matchlength:], value)
+		var new node
+		matchlength := MatchingNibbleLength(key, n.Key)
+		if matchlength == len(n.Key) {
+			new = self.insert(n.Val, key[matchlength:], value)
 		} else {
-			pnode := self.insert(nil, k[matchlength+1:], cnode)
+			pnode := self.insert(nil, n.Key[matchlength+1:], n.Val)
 			nnode := self.insert(nil, key[matchlength+1:], value)
-			fulln := NewFullNode(self)
-			fulln.dirty = true
-			fulln.set(k[matchlength], pnode)
-			fulln.set(key[matchlength], nnode)
-			n = fulln
+			var full fullNode
+			full[n.Key[matchlength]] = pnode
+			full[key[matchlength]] = nnode
+			new = full
 		}
 		if matchlength == 0 {
-			return n
+			return new
 		}
+		return shortNode{key[:matchlength], new}
 
-		snode := NewShortNode(self, key[:matchlength], n)
-		snode.dirty = true
-		return snode
-
-	case *FullNode:
-		cpy := node.Copy(self).(*FullNode)
-		cpy.set(key[0], self.insert(node.branch(key[0]), key[1:], value))
-		cpy.dirty = true
+	case fullNode:
+		n[key[0]] = self.insert(n[key[0]], key[1:], value)
+		return n
 
-		return cpy
+	case hashNode:
+		return self.insert(self.resolveHash(n), key, value)
 
 	default:
-		panic(fmt.Sprintf("%T: invalid node: %v", node, node))
+		panic(fmt.Sprintf("%T: invalid node: %v", n, n))
 	}
 }
 
-func (self *Trie) get(node Node, key []byte) Node {
-	if len(key) == 0 {
-		return node
-	}
-
-	if node == nil {
-		return nil
-	}
-
-	switch node := node.(type) {
-	case *ShortNode:
-		k := node.Key()
-		cnode := node.Value()
-
-		if len(key) >= len(k) && bytes.Equal(k, key[:len(k)]) {
-			return self.get(cnode, key[len(k):])
-		}
-
-		return nil
-	case *FullNode:
-		return self.get(node.branch(key[0]), key[1:])
-	default:
-		panic(fmt.Sprintf("%T: invalid node: %v", node, node))
-	}
+func concat(s1 []byte, s2 ...byte) []byte {
```

Comment:
If only there was a way to indicate that this PR is in progress. Maybe a label of some sort. Someone should invent that. append‎, being nice and all, is allowed to reuse the backing array of its first argument if it has sufficient cap. concat doesn't reuse the first argument but returns a fresh slice every time. I'll leave it as an exercise for you to puzzle about why concat is being used instead of append in the one place where it is used. 